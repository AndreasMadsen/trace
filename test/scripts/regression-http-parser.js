'use strict';

const http = require('http');

//
// Test an issue where the async context from HTTP parser isn't correctly
// captured, because MakeCallback is called synchronously.
//

const server = http.createServer(function (req, res) {
  res.end('hallo world');
});

server.listen(0, 'localhost', function () {
  const addr = server.address();
  http.get(`http://${addr.address}:${addr.port}`, function (res) {

    // there is no public endpoint where we lose context. However
    // the internal _read that is called from TCP.onread -> socket.push
    // does lose context.
    // An alternative is also to overwrite res.socket.read, but that
    // is perhaps less public.
    const _read = res.socket._read;
    res.socket._read = function () {
      throw new Error('trace');
      return _read.apply(this, arguments);
    };

    res.resume();
    res.once('end', server.close.bind(server));
  });
});
