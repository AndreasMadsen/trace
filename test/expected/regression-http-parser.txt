/scripts/regression-http-parser.js:25
      throw new Error('trace');
      ^

Error: trace
    at Socket.res.socket._read (/scripts/regression-http-parser.js:25:13)
    at ____________________:0:0
    at Server.<anonymous> (/scripts/regression-http-parser.js:16:8)
    at ____________________:0:0
    at ____________________:0:0
    at ____________________:0:0
    at Object.<anonymous> (/scripts/regression-http-parser.js:14:8)
